### PR TITLE
MS-1017 Cookie management for IE

### PR DESCRIPTION
**Pull request by valentyng**

### CODE REVIEW

**Comment by valentyng** [comment]
Don't merge yet

File: ie/source/forge/NativeExtensions.cpp [hunk 1]
```diff
@@ -341,3 +342,112 @@ STDMETHODIMP CNativeExtensions::notification(BSTR icon, BSTR title, BSTR text,
...
+{
+    logger->debug(L"NativeExtensions::cookies_get "
+        L" -> " + wstring(url) +
+        L" -> " + wstring(name));
+
+
+    if (this->tabId == 0) {
+        // it's a magic. don't touch it!
+        TCHAR szURL[256] = { 0 };
+        _tcscpy(szURL, url);
```
Comment (blacktaxi) [inline ie/source/forge/NativeExtensions.cpp]: How will this behave if `url` is longer than 256 `TCHAR`s?

File: ie/source/forge/NativeExtensions.cpp [hunk 2]
```diff
@@ -341,3 +342,111 @@ STDMETHODIMP CNativeExtensions::notification(BSTR icon, BSTR title, BSTR text,
...
+            if (GetLastError() == ERROR_INSUFFICIENT_BUFFER)
+            {
+                delete[]cookieData;
+                cookieData = new TCHAR[dwSize + 1];
+                memset(cookieData, 0, sizeof(TCHAR)*(dwSize + 1));
+                // Try the call again.
+                if (InternetGetCookieEx(szURL, cookieName, cookieData, &dwSize, 0, NULL))
+                {
+                    CComDispatchDriver(success).Invoke1((DISPID)0, &CComVariant(wstring(cookieData).c_str()));
+                    delete[] cookieData;
```
Comment (blacktaxi) [inline ie/source/forge/NativeExtensions.cpp]: Is it ok to free this memory right after calling the `success` handler? What would be an example of the `success` handler?

File: ie/source/forge/NativeExtensions.cpp [hunk 3]
```diff
@@ -341,3 +342,111 @@ STDMETHODIMP CNativeExtensions::notification(BSTR icon, BSTR title, BSTR text,
...
+                delete[] cookieData;
+            }
+            else
+            {
+                logger->error(L"NativeExtensions::cookies_get failed"
+                    L" -> " + wstring(url) +
+                    L" -> " + wstring(name));
+
+                wstring message = L"Error code: " + std::to_wstring(GetLastError());
+                CComDispatchDriver(error).Invoke1((DISPID)0, &CComVariant(message.c_str()));
```
Comment (blacktaxi) [inline ie/source/forge/NativeExtensions.cpp]: For example, here we don't free the memory allocated for `message`. Is this inconsistency intentional?

**Comment by valentyng** [comment]
@blacktaxi, I've tested - it works.

**Pull request merged**